doc/mancha.1
.\" Automatically generated by Pandoc 3.9.0.2
.\"
.TH "MANCHA" "1"
.SH NAME
mancha \- view manual pages via cha(1)
.SH SYNOPSIS
\f[B]mancha\f[R] [\f[B]\-M\f[R] \f[I]path\f[R]] [\f[I]section\f[R]]
\f[I]name\f[R]
.PD 0
.P
.PD
\f[B]mancha\f[R] [\f[B]\-M\f[R] \f[I]path\f[R]] [\f[I]section\f[R]]
\f[B]\-k\f[R] \f[I]keyword\f[R]
.PD 0
.P
.PD
\f[B]mancha\f[R] \f[B]\-l\f[R] \f[I]file\f[R]
.SH DESCRIPTION
\f[B]mancha\f[R] enables viewing man pages using the Chawan browser.
It is analogous to the \f[B]w3mman\f[R](1) utility.
.PP
\f[B]mancha\f[R] executes \f[B]cha\f[R](1) with the appropriate
\f[I]man:\f[R], \f[I]man\-k:\f[R] or \f[I]man\-l:\f[R] URLs.
The protocol adapter then opens the man page and injects markup into it,
e.g.\ man page references are converted into \f[I]man:\f[R] links.
.SH OPTIONS
Command line options are:
.TP
\f[B]\-M\f[R] \f[I]path\f[R]
Set \f[I]path\f[R] as the MANPATH environment variable.
See \f[B]man\f[R](1) for details of how this is interpreted.
.TP
\f[B]\-k\f[R] \f[I]keyword\f[R]
Use \f[I]keyword\f[R] for keyword\-based man page search.
.TP
\f[B]\-l\f[R] \f[I]file\f[R]
Open the specified local \f[I]file\f[R] as a man page.
.SH ENVIRONMENT
Following environment variables are used:
.TP
\f[B]MANCHA_CHA\f[R]
If set, the contents of the variable are used instead of \f[I]cha\f[R].
(Note that the \f[I]cha\f[R] command is called through
\f[B]system\f[R](3), so you do not have to override it so long as
\f[I]cha\f[R] is found in your \f[B]PATH\f[R].)
.TP
\f[B]MANCHA_MAN\f[R]
If set, the contents of the variable are used instead of
\f[I]/usr/bin/man\f[R].
.TP
\f[B]MANCHA_APROPOS\f[R]
If set, the contents of the variable are used instead of
\f[I]/usr/bin/man\f[R].
.RS
.PP
(This is not a typo; normally (except on FreeBSD), \f[B]mancha\f[R]
assumes that \f[B]man\f[R](1) is compatible with \f[B]apropos\f[R](1)
and accepts the \f[I]\-s\f[R] parameter.
Overriding \f[B]MANCHA_MAN\f[R] therefore also overrides the command
used for \f[B]man\-k\f[R], so long as \f[B]MANCHA_APROPOS\f[R] is not
set.)
.RE
.SH SEE ALSO
\f[B]man\f[R](1), \f[B]cha\f[R](1), \f[B]cha\-cgi\f[R](5),
\f[B]w3mman\f[R](1)


NEWS
unreleased

* dupeBuffer/anchor navigation no longer forks buffer process
* support VT420
* support DEC special graphics in ASCII mode
* support all VP8L indexed color transforms
* support pandoc-style definition list in markdown
* support color, width, align, noshade properties on HR element
* support Element#getAttributeNames
* add status.format-mode (sets formatting of status bar)
* add toggleLinkHintsAutoClick command (by Arthur Jacquin)
* add saveScreen, editScreen commands; rename sourceEdit to editSource
* add x-type, x-match, x-nc-match mailcap extensions (from w3mmee)
* improve keybinding format (mostly backwards-compatible; see cha-config(5))
* remove display.force-clear (shouldn't have ever been needed)
* remove display.query-da1 (no longer needed)
* remove deprecated special case in pager.load() with trailing newline
* remove deprecated parentBuffer, prevSiblingBuffer, nextSiblingBuffer actions
* make terminal querying asynchronous
* work around FreeBSD vt(4) freezing on startup
* optimize out Sixel repaints when scrolling
* drop excessive frames when the terminal/link cannot keep up
* rework seccomp handling to support cross-compilation to incompatible archs
* fix cached images not being cleared when a buffer is deleted
* fix a NULL deref in jebp
* fix y/n being truncated out of long ask() prompts
* fix media queries not being re-evaluated on windowChange
* fix load info not being set on cycling between buffers
* fix color blending over body bgcolor
* fix matching of nested subsequent sibling combinator
* fix various CSS image sizing bugs
* fix percentage calc() output in color values
* fix inline pre/nowrap not being flushed to the next line on overflow
* fix bugs in our libregexp UTF-8 patch
* fix various proxy authentication bugs
* fix [attr|=value] and :lang() pseudo-class bugs
* fix localStorage.key function
* fix gotoLine malfunctioning in select menu
* fix bug in negation of the minimum layout unit value
* fix crash on Storage#key overflow on 32-bit targets
* fix crash on calling status getter for unsent XHR
* fix various charset decoder bugs
* fix TypedArray handling
* fix suspend() not suspending all processes

0.3.0 (2025.11.21)

UI additions
* wheel-scroll, side-wheel-scroll config options (by Tim Culverhouse)
* new API functions Pager#traverse, Pager#navDirection, Pager#revDirection,
  Buffer#find
* bracketed paste - pasting a URL now opens the URL bar automatically
* "copy URL", "copy selection" now works over SSH when OSC 52 is supported
* "hints mode" for jumping to specific links on the page (default
  keybinding is `f`, configurable using `input.link-hint-chars`)
* added loaded stylesheet/image counters
---
CSS additions
* monochrome, color, color-index media queries
* implement <ol start=..., <li value=...
* -cha-content-type media feature to detect content type of current page
* complete var() support, improved calc() support, @layer
* support CSS borders
* support progress element (rendering inspired by w3m's download panel)
* support non-UTF-8 CSS files
---
JS additions
* NodeIterator, AbortSignal (etc.), onsubmit
* getBoundingClientRect, getClientRects (only returns real values in app mode)
* better module script support
---
Misc additions
* support blockquote, reference links, block lists in markdown
* support OSC 8 in ANSI text parser
* support more kinds of webp images (VP8X, VP8 with coeff. skipping, etc.)
* support customizing gopher selector text by moving it to ua.css
---
UI changes
* improved highlighting in monochrome mode
* more sensible semantics for pager.load() - empty string now works, newline at
  end deprecated in favor of pager.loadSubmit()
* mouse improvements:
	- click on status bar opens URL bar: left click -> current URL, right
	  click -> URL under cursor, middle click -> empty URL bar
	- double click + drag selects text
* fixed flash of unstyled content and removed blank screen on page load
* moved all commands in `cmd.pager` and `cmd.buffer` into `cmd`
* binding keys no longer requires `cmd.` prefix for built-in commands
* color contrast correction is no longer invoked on cells that do not use
  the terminal theme for foreground or background
---
Misc changes
* history, cookies, etc. are now saved in XDG_DATA_HOME for XDG basedir users
  (instead of XDG_CONFIG_HOME)
	- users that store their config in ~/.chawan are not affected
* unify binaries that link to OpenSSL in dynamically linked builds, and
  unify markdown, gemini, etc. -> HTML converters
* allow overriding pkg-config path in Makefile (using PKG_CONFIG)
* better color approximation in "eight-bit" color mode (256 colors)
* #L{n} (e.g. #L1) jumps to nth line in plain text files (idea from Dillo)
* console buffer is now spawned on demand
* switched from QJS-NG to QJS (with patches from NG)
---
Removals
* remove deprecated prepend-https, css and stylesheet config options (use
  prepend-scheme, user-style instead)
* remove protocol.form-request config option (read the body in CGI instead)
* remove buffer tree; buffers are now organized in lists, and buffer lists
  belong in tabs (note: the latter does not have a real UI yet)
---
Optimizations
* layout caching (makes a huge difference in memory usage and performance)
* CSS parser is now single-pass (kind of)
* cascade: eliminate several allocations, merge hash maps for separate origins
* more compact DOM representation (uses doubly linked list instead of a vector)
* insert style sheets incrementally instead of traversing DOM several times
* async terminal handling (printing large Sixels no longer blocks the rest of
  the pager)
---
Fixes
* fix buffer cloning on musl
* fix crash when saving cookies
* fix possible crash in CanvasRenderingContext2D#closePath
* fix mystery miscompilation bug on macOS
* fix interactive mode when reading from stdin on macOS
* fix referrer policy handling with referer-from=true
* fix compilation on OpenBSD + ancient Nim
* fix crash on hitting C-c on buffers waiting for connection

adapter/tools/mancha.nim
{.push raises: [].}

import std/os

import io/chafile
import types/opt
import utils/myposix
import utils/twtstr

proc help(i: int) =
    let s = """
Usage:
mancha [-M path] [[-s] section] -k keyword
mancha [-M path] [[-s] section] name
mancha -l file"""
    if i == 0:
      discard cast[ChaFile](stdout).writeLine(s)
    else:
      discard cast[ChaFile](stderr).writeLine(s)
    quit(i)

var i = 1
let n = paramCount()

proc getnext(): string =
  inc i
  if i <= n:
    return paramStr(i)
  help(1)
  ""

proc main() =
  let n = paramCount()
  var section = ""
  var local = ""
  var man = ""
  var keyword = ""
  var forceSection = false
  while i <= n:
    let s = paramStr(i)
    if s == "":
      inc i
      continue
    let c = s[0]
    let L = s.len
    if c == '-':
      if s.len != 2:
        help(1)
      case s[1]
      of 'h': help(0)
      of 'M':
        if twtstr.setEnv("MANPATH", getnext()).isErr:
          discard cast[ChaFile](stderr).writeLine("Failed to set MANPATH")
          quit(1)
      of 's':
        section = getnext()
        forceSection = true
      of 'l': local = getnext()
      of 'k': keyword = getnext()
      else: help(1)
    elif section == "" and (c in {'0'..'9'} or L == 1 and c in {'n', 'l', 'x'}):
      section = s
    elif man == "":
      man = s
    else:
      help(1)
    inc i
  if not forceSection and section != "" and man == "" and keyword == "" and
      local == "":
    man = section
    section = ""
  if not ((local != "") != (man != "") != (keyword != "")):
    help(1)
  if local != "" and section != "":
    help(1)
  let qsec = if section != "": '(' & section & ')' else: ""
  let query = if local != "":
    if local[0] == '~':
      local = expandTilde(local)
    elif local[0] != '/':
      local = (myposix.getcwd() / local)
    "man-l:" & local
  elif keyword != "":
    "man-k:" & keyword & qsec
  else:
    "man:" & man & qsec
  var cha = getEnvEmpty("MANCHA_CHA")
  if cha == "":
    cha = "cha"
  let cmd = cha & " " & quoteShellPosix(query)
  let res = myposix.system(cstring(cmd))
  if res < 0:
    discard cast[ChaFile](stderr).writeLine("failed to fork child process")
    quit(1)
  quit(exitStatusLikeShell(res))

main()

{.pop.} # raises: []
